full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,155 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+/* eslint-disable @typescript-eslint/no-empty-function -- Remove me */
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    const allTodos = async () => {
+      try {
+        const result = await fetch('/api/todos');
+        if (!result.ok) throw new Error('fetch for todos failed');
+        const formattedResult = await result.json();
+        setTodos(formattedResult);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    };
+    allTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const todoAddResult = await fetch('/api/todos', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(newTodo),
+      });
+      if (!todoAddResult.ok) throw new Error('add todo failed');
+      const formattedResult = (await todoAddResult.json()) as Todo;
+      const todosCopy = todos.slice();
+      todosCopy.push(formattedResult);
```

Comment:
This code is good. Personally, I prefer this:
`setTodos([...todos, formattedResult]);`